### PR TITLE
Fix reclaim and preempt not conside min resource

### PR DESCRIPTION
**Pull request by zbbkeepgoing**
fix: #2740

### CODE REVIEW

File: pkg/scheduler/plugins/gang/gang.go [hunk 1]
```diff
@@ -83,15 +83,26 @@ func (gp *gangPlugin) OnSessionOpen(ssn *framework.Session) {
...
 			job := ssn.Jobs[preemptee.Job]
 			if _, found := jobOccupiedMap[job.UID]; !found {
 				jobOccupiedMap[job.UID] = job.ReadyTaskNum()
+				jobAllocatedMap[job.UID] = util.GetAllocatedResource(job)
+			}
+
+			// It's not allowed to reclaim or preempt task when allocated is less equal to min resource of job.
+			jobMinResource := job.GetMinResources()
+			klog.V(4).Infof("Job name is <%v>, Allocated is %+v, MinResource is %+v.", job.Name, jobAllocatedMap[job.UID], jobMinResource)
+			if jobAllocatedMap[job.UID].LessEqualPartly(jobMinResource, api.Zero) {
```
Comment (wangyang0616) [inline pkg/scheduler/plugins/gang/gang.go:97]: Before judging with jobMinAvailble, is it more reasonable to subtract the resource information of preemptee from the allocated resources?

**Comment by wangyang0616** [comment]
reviewed
cc @jiangkaihua @william-wang

**Comment by zbbkeepgoing** [comment]
keep active

**Review by hwdef — COMMENTED** [review]
/lgtm
I agree with this.

File: pkg/scheduler/plugins/gang/gang.go [hunk 2]
```diff
@@ -83,15 +83,27 @@ func (gp *gangPlugin) OnSessionOpen(ssn *framework.Session) {
 	preemptableFn := func(preemptor *api.TaskInfo, preemptees []*api.TaskInfo) ([]*api.TaskInfo, int) {
 		var victims []*api.TaskInfo
 		jobOccupiedMap := map[api.JobID]int32{}
-
+		jobAllocatedMap := map[api.JobID]*api.Resource{}
 		for _, preemptee := range preemptees {
 			job := ssn.Jobs[preemptee.Job]
 			if _, found := jobOccupiedMap[job.UID]; !found {
 				jobOccupiedMap[job.UID] = job.ReadyTaskNum()
+				jobAllocatedMap[job.UID] = util.GetAllocatedResource(job)
```
Comment (lowang-bh) [inline pkg/scheduler/plugins/gang/gang.go:91]: It is better to clone from job's allocated.

File: pkg/scheduler/plugins/gang/gang.go [hunk 3]
```diff
@@ -83,15 +83,27 @@ func (gp *gangPlugin) OnSessionOpen(ssn *framework.Session) {
...
+			jobAllocatedTmp := jobAllocatedMap[job.UID].Clone()
+			if jobAllocatedTmp.Sub(preemptee.Resreq).LessPartly(jobMinResource, api.Zero) {
+				klog.V(4).Infof("Can not preempt or reclaim task <%v/%v>, because allocated of job <%v> is less equal to min resource.",
+					preemptee.Namespace, preemptee.Name, job.Name)
+				continue
 			}
 
 			if jobOccupiedMap[job.UID] > job.MinAvailable {
 				jobOccupiedMap[job.UID]--
+				jobAllocatedMap[job.UID].Sub(preemptee.Resreq)
```
Comment (lowang-bh) [inline pkg/scheduler/plugins/gang/gang.go:106]: Here, two sub callings maybe can be merged to once.

**Pull request closed without merging**